server/base/src/main/java/org/apache/accumulo/server/util/ratelimit/SharedRateLimiterFactory.java

```diff
@@ -0,0 +1,152 @@
+
+        // Report periodically
+        timer.schedule(new Runnable() {
+          @Override
+          public void run() {
+            instance.report();
+          }
+        }, REPORT_RATE, REPORT_RATE);
+      }
+      return instance;
+    }
+  }
+
+  public static SharedRateLimiterFactory getInstance(AccumuloConfiguration conf) {
+    return getInstance(SimpleTimer.getInstance(conf));
+  }
+
+  /**
+   * Lookup the RateLimiter associated with the specified name, or create a new one for that name. RateLimiters should be closed when no longer needed.
+   *
+   * @param name
+   *          key for the rate limiter
+   * @param rateGenerator
+   *          a function which can be called to get what the current rate for the rate limiter should be.
+   */
+  public RateLimiter create(String name, Callable<Long> rateGenerator) {
+    synchronized (activeLimiters) {
+      if (activeLimiters.containsKey(name)) {
+        SharedRateLimiter limiter = activeLimiters.get(name);
+        return limiter;
+      } else {
+        long initialRate;
+        try {
+          initialRate = rateGenerator.call();
+        } catch (Exception ex) {
+          throw new IllegalStateException(ex);
+        }
+        SharedRateLimiter limiter = new SharedRateLimiter(name, rateGenerator, initialRate);
+        activeLimiters.put(name, limiter);
+        return limiter;
+      }
+    }
+  }
+
+  protected void update() {
+    List<SharedRateLimiter> limiters;
+    synchronized (activeLimiters) {
+      limiters = ImmutableList.copyOf(activeLimiters.values());
+    }
+    for (SharedRateLimiter limiter : limiters) {
+      limiter.update();
+    }
+  }
+
+  protected void report() {
+    List<SharedRateLimiter> limiters;
+    synchronized (activeLimiters) {
+      limiters = ImmutableList.copyOf(activeLimiters.values());
+    }
+    for (SharedRateLimiter limiter : limiters) {
+      limiter.report();
+    }
+  }
+
+  protected class SharedRateLimiter extends GuavaRateLimiter {
+    private final LongAdder permitsAcquired = new LongAdder();
+    private final Callable<Long> rateCallable;
+    private final String name;
+
+    SharedRateLimiter(String name, Callable<Long> rateCallable, long initialRate) {
+      super(initialRate);
+      this.name = name;
+      this.rateCallable = rateCallable;
+    }
+
+    @Override
+    public void acquire(long permits) {
+      super.acquire(permits);
+      permitsAcquired.add(permits);
+    }
+
+    public void update() {
+      try {
+        // Reset rate if needed
+        long rate = rateCallable.call();
+        if (rate != getRate()) {
+          setRate(rate);
+        }
+      } catch (Exception ex) {
+        logger.debug("Failed to update rate limiter", ex);
+      }
+    }
+
+    public void report() {
+      long sum = permitsAcquired.sumThenReset();
+      if (sum > 0) {
+        logger.debug(String.format("RateLimiter '%s': %,d of %,d permits/second", name, sum * 1000L / REPORT_RATE, getRate()));
```

Comment:
Can the logger configuration be changed at runtime?  It might make sense simply not to schedule the reporting task if debug logging is disabled at startup.